
# Лабораторная работа 4

## Выполнил: Морозов Алексей, М32071

## Вариант: 9, задание: 5


In [56]:
import numpy as np
import scipy as scp
import matplotlib as mlt
import matplotlib.pyplot as plt
import pandas as pd

import matplotlib.axes._axes as axes
import matplotlib.figure as figure

В файле $MEN\_SHOES.csv$ приведены данные о продажах мужской обуви. Для каждого вопроса требуется формализовать задачу и проверить статистическую гипотезу двумя критериями, если не сказано иное

In [57]:
data = pd.read_csv('MEN_SHOES.csv', thousands=',')
data['Current_Price'] = data['Current_Price'].str.replace('₹', '', regex=True).replace(',', '', regex=True).astype(float)
count = len(data)
data

Brand_Name  How_Many_Sold  Current_Price  \
0          ASIAN           2242         1098.0   
1          ASIAN            240          674.0   
2          ASIAN          16662          588.0   
3          ASIAN            135          599.0   
4         Reebok            240          982.0   
...          ...            ...            ...   
23935     Campus            198          623.0   
23936  Wakefield             49          399.0   
23937       URJO            557          839.0   
23938    FEETEES             63          349.0   
23939       FURO              2         1299.0   

                                         Product_details  RATING  
0      Oxygen-01 Sports Running,Walking & Gym Shoes w...     3.8  
1      Men's Express-08 Sports Running,Walking,Gym,Tr...     4.0  
2      Men's Cosko Sports Running,Walking,Gym,Trainin...     3.8  
3      Wind-03 Sports Running,Walking & Gym Shoes wit...     4.0  
4                Men's Velocity Runner Lp Running Shoe       4.0  
...                                                  ...     ...  
23935                   Men's HARVEL PRO Running Shoes       4.0  
23936        Stylish Running Sports Shoes for MEN-2010       3.4  
23937                               Mens Running Shoes       4.2  
23938  Nexon Men's Casual Eva Socks Knitted Running S...     3.6  
23939  Low Ankle Running Sports Shoes for Men (O-5033)       3.0  

[23940 rows x 5 columns]

Для проверки гипотез буду использовать уровень значимости $\alpha = 0.05$

In [58]:
alpha = 0.05


##  Задача 1

### Условие

Разумно ли количество проданных экземпляров обуви аппроксимировать распределением Пуассона, а рейтинг – нормальным распределением (по 1 тесту на каждый вопрос)?



### Решение

Для первого вопроса возьмем за нулевую гипотезу о том, что количество проданных экземпляров обуви имеет распределение Пуассона, тогда первая будет ее отрицанием

\begin{gather}
    H_0 = \text{количество проданных экземпляров обуви имеет распределение Пуассона} \\
    H_1 = \neg{H_0}
\end{gather}

Используем критерий согласия хи-квадрат, соответствующая статистика имеет вид

\begin{equation}
    \chi^2 = N\sum{\frac{(O_i - E_i)^2}{E_i}}
\end{equation}

где
* $O_i$ - наблюдаемая частота,
* $E_i$ - ожидаемая при истинности нулевой гипотезы частота $i$-го наблюдения


In [59]:
sold_ser = data.How_Many_Sold
sold_mean = sold_ser.mean()
x_tmp, counts_tmp = np.unique(sold_ser, return_counts=True)

chi2_sold_df = pd.DataFrame({'x': x_tmp, 'n': counts_tmp})
chi2_sold_df['n_exp'] = len(data) * scp.stats.poisson.pmf(k=chi2_sold_df.x, mu=sold_mean)
chi2_sold_df

x     n          n_exp
0       2   399   0.000000e+00
1       7   399   0.000000e+00
2      12   399   0.000000e+00
3      49   399   0.000000e+00
4      60   399   0.000000e+00
5      63   399   0.000000e+00
6      83   399   0.000000e+00
7      84   399   0.000000e+00
8     102   399   0.000000e+00
9     112   798   0.000000e+00
10    116   399   0.000000e+00
11    135   399   0.000000e+00
12    146   399   0.000000e+00
13    165   399   0.000000e+00
14    173   399   0.000000e+00
15    198   399   0.000000e+00
16    206   399   0.000000e+00
17    216   399   0.000000e+00
18    240  1197   0.000000e+00
19    245   399   0.000000e+00
20    277   399   0.000000e+00
21    307   399   0.000000e+00
22    309   399   0.000000e+00
23    317   399   0.000000e+00
24    350   399   0.000000e+00
25    382   399   0.000000e+00
26    392   399   0.000000e+00
27    421   399   0.000000e+00
28    455   399   0.000000e+00
29    557   399   0.000000e+00
30    658   399   0.000000e+00
31    682   399   0.000000e+00
32    706   399   0.000000e+00
33    812   399   0.000000e+00
34    843   399   0.000000e+00
35    860   399   0.000000e+00
36    954   399   0.000000e+00
37    998   399   0.000000e+00
38   1060   399   0.000000e+00
39   1072   399   0.000000e+00
40   1490   399  6.387083e-318
41   1685   399  5.160108e-251
42   1795   399  1.132729e-217
43   2076   399  2.100843e-145
44   2144   399  1.727720e-130
45   2242   399  1.237972e-110
46   3448   399   1.150643e+02
47   3532   399   1.350707e+02
48   4583   399   3.249261e-65
49   4992   399  5.734840e-121
50   5690   399  3.735233e-249
51   6159   399   0.000000e+00
52   7712   399   0.000000e+00
53  16662   399   0.000000e+00
54  18747   399   0.000000e+00
55  38839   399   0.000000e+00
56  72611   399   0.000000e+00

In [60]:
stat = np.sum(((chi2_sold_df.n - chi2_sold_df.n_exp)**2) / chi2_sold_df.n_exp)
print(f'Вычисленная статистика: {stat}')

Вычисленная статистика: inf


Распределение Пуассона зависит от одного параметра, поэтому статистика имеет распределение хи-квадрат с $N - 2$ степенями свободы. Для вычисления критического значения статистики используется квантиль $1 - \alpha$ (как и для остальных критериев)

In [61]:
print(f'Критическое значение: {scp.stats.chi2.isf(alpha, len(chi2_sold_df) - 2)}')

Критическое значение: 73.31149302908324


Вычисленная статистика больше критического значения при уровне значимости $\alpha$, а значит нулевая гипотеза при выбранном уровне значимости отвергается. Следовательно, аппроксимировать количество проданной обуви распределением Пуассона нельзя

Для второго вопроса возьмем за нулевую гипотезу предположение о том, что рейтинг имеет нормальное распределение с параметрами, вычисленными методом максимального правдоподобия

\begin{gather}
    H_0 = \text{рейтинг имеет нормальное распределение} \\
    H_1 = \neg{H_0}
\end{gather}

Используем критерий Колмогорова-Симпсона-Лиллиефорса
Статистика критерия вычисляется аналогично статистике критерия Колмогорова-Смирнова:

\begin{equation}
    D_n = \max{\left(\max{\left(\frac{i}{N} - \Phi\left(\frac{X_{(i)} - \bar{X}}{s^2}\right)\right)}, \max{\left(\Phi\left(\frac{X_{(i)} - \bar{X}}{s^2}\right) - \frac{i - 1}{N}\right)}\right)}
\end{equation}

для всех $1 \leq i \leq N$, где
* $\bar{X}$ - выборочная медиана,
* $s^2$ - выборочная несмещенная дисперсия,
* $X_{(i)}$ - $i$-ый элемент вариационного ряда

In [62]:
rating_arr = data['RATING'].values
rating_mean = rating_arr.mean()
rating_var = rating_arr.var()

rating_arr.sort()
stat = max([max(abs((i + 1) / count) - scp.stats.norm.cdf((rating_arr[i] - rating_mean) / np.sqrt(rating_var)), scp.stats.norm.cdf((rating_arr[i] - rating_mean) / np.sqrt(rating_var)) - i / count) for i in range(count)])

print(f'Вычисленная статистика: {stat}')

Вычисленная статистика: 0.23791404729168092


Критическое значение вычисляется по таблице

In [63]:
print(f'Критическое значение: {0.895/((0.83 + count) / np.sqrt(count) - 0.01)}')

Критическое значение: 0.005784608536943777


Критическое значение меньше вычисленного значения статистики, а значит нулевая гипотеза отвергается и рейтинг не имеет нормального распределения


## Задача 2

### Условие

Верно ли что распределения количества проданных экземпляров существенно не отличаются в зависимости от бренда? Тот же вопрос для цены (по одному тесту на утверждение).



### Решение

Нулевой гипотезой в первом вопросе будет гипотеза, что распределение количества проданной обуви одинаково для всех брендов

\begin{gather}
    H_0 = \text{распределение продаж одинаково для всех брендов} \\
    H_1 = \neg{H_0}
\end{gather}

Используем критерий Крускала-Уоллеса

Сначала необходимо проранжировать все данные вместе (каждому значению присвоить ранг по всей выборке), затем разбить данные с рангами по группам и вычислить статистику

\begin{equation}
    H = (N - 1)\frac{\sum_{i=1}^g{n_i\left(\bar{r_i} - \bar{r}\right)^2}}{\sum_{i=1}^{g}{\sum_{j=1}^{n_i}{\left(r_{ij} - \bar{r}\right)^2}}}
\end{equation}

где
* $N$ - суммарный объем выборок,
* $g$ - число групп,
* $n_i$ - число наблюдений в $i$-ой группе,
* $r_{ij}$ - ранг $j$-ого наблюдения в $i$-ой группе,
* $\bar{R_i}$ - средний ранг в группе $i$,
* $\bar{r}$ - средний ранг из всех наблюдений

In [64]:
sold_ranks = scp.stats.rankdata(data.How_Many_Sold)
data_sold_ranks = data.assign(rank_sold=sold_ranks)
unique_brands_ranked_df_arr = [data_sold_ranks[data_sold_ranks['Brand_Name'] == name] for name in data_sold_ranks['Brand_Name'].unique()]
dividend = divisor = 0
for i in range(len(unique_brands_ranked_df_arr)):
    brand_df = unique_brands_ranked_df_arr[i]
    dividend += len(brand_df) * ((brand_df.rank_sold.mean() - (len(data_sold_ranks) - 1) / 2)**2)
    for j in range(len(brand_df)):
        divisor += (brand_df.iloc[i].rank_sold - (len(data_sold_ranks) - 1) / 2)**2
stat = (len(data_sold_ranks) - 1) * dividend / divisor
print(f'Вычисленная статистика: {stat}')

Вычисленная статистика: 7904.696328668118


Статистика $H$ подчиняется распределению хи-квадрат с $g - 1$ степенями свободы

In [65]:
print(f'Критическое значение: {scp.stats.chi2.isf(alpha, df=(len(unique_brands_ranked_df_arr) - 1))}')

Критическое значение: 31.41043284423092


Критическое значение меньше вычисленного значения, поэтому нулевая гипотеза отвергается: распределение продаж от бренда к бренду различается

Во втором вопросе нулевой будет гипотеза, что распределение цены одинаково для всех брендов

\begin{gather}
    H_0 = \text{распределение цен одинаково для всех брендов} \\
    H_1 = \neg{H_0}
\end{gather}

Используем критерий хи-квадрат со статистикой

\begin{equation}
    \chi^2 = \sum_{i=1}^{g}{\sum_{j=S_1}^{S_l}{\frac{\left(\nu_{ij} - n_i p^{(j)}\right)^2}{n_{i} p^{(j)}}}}
\end{equation}

где
* $N$ - суммарный объем выборки
* $S$ - набор полученных значений размером $l$,
* $g$ - число групп,
* $\nu_{ij}$ - частота значения $j$ в $i$-ой группе,
* $n_i$ - объем $i$-ой группы
* $p^{(j)} = \frac{\sum_{i=1}^{g}{\nu_{ij}}}{N}$ - вероятность значения $j$ среди всей выборки

In [66]:
unique_brands_df_arr = [data[data['Brand_Name'] == name] for name in data['Brand_Name'].unique()]
price_freq_df_arr = [pd.DataFrame(np.unique(unique_brands_df_arr[i].Current_Price, return_counts=True)).transpose().rename(columns={0: 'price', 1: 'freq'}) for i in range(len(unique_brands_df_arr))]
stat = 0
for i in range(len(price_freq_df_arr)):
    brand_df = price_freq_df_arr[i]
    for price in data['Current_Price'].unique():
        if not np.isnan(price):
            stat += (np.sum(brand_df[brand_df.price == price].freq) - brand_df.freq.sum() * (len(data[data.Current_Price == price]) / len(data)))**2 / (brand_df.freq.sum() * (len(data[data.Current_Price == price]) / len(data)))
        else:
            stat += (np.sum(brand_df[brand_df.price.isna()].freq) - brand_df.freq.sum() * (len(data[data.Current_Price.isna()]) / len(data)))**2 / (brand_df.freq.sum() * (len(data[data.Current_Price.isna()]) / len(data)))
print(f'Значение статистики: {stat}')

Значение статистики: 413915.0000000054


Статистика подчиняется распределению хи-квадрат с $(N - 1)(g - 1)$ степенями свободы

In [67]:
print(f'Критическое значение: {scp.stats.chi2.isf(alpha, df=((len(data.Current_Price.unique()) - 1) * len(price_freq_df_arr)))}')

Критическое значение: 1082.9731737565853


Статистика больше критического значения, поэтому гипотеза, что распределение цен для всех брендов одинаково отвергается


## Задача 3

### Условие

Есть подозрение, что рейтинг зависит от цены. Проверить данное предположение


### Решение

За нулевую гипотезу возьму предположение, что рейтинг и цена независимы, альтернативой будет обратное - что они зависимы

 \begin{gather}
    H_0 = \text{рейтинг не зависит от цены} \\
    H_1 = \neg{H_0}
\end{gather}

Для корректности работы критериев буду оценивать независимость только тех данных, в которых нет NaN-значений, отклонив $798$ строк (около $3\%$ значений)

In [68]:
rating_ser = data[~data['Current_Price'].isna()].RATING #type: pd.Series
sold_ser = data[~data['Current_Price'].isna()].Current_Price #type: pd.Series

rating_ser_mean = rating_ser.mean()
sold_ser_mean = sold_ser.mean()

rating_ser_var = rating_ser.var()
sold_ser_var = sold_ser.var()

Первым критерием будет критерий, основанный на коэффициенте корреляции Пирсона

\begin{equation}
    \rho = \frac{1}{N - 1} \frac{\sum_{i=1}^{N}{\left(x_i - \bar{x}\right)\left(y_i - \bar{y}\right)}}{\sqrt{s^2_x s^2_y}}
\end{equation}

где
* $x$, $y$ - выборки размером $N$
* $\bar{x}$, $s^2_x$ и $\bar{y}$, $s^2_y$ - медиана и среднее соответственно $x$ и $y$

In [69]:
pearson_corr = (1 / (len(rating_ser) - 1) * np.sum((rating_ser - rating_ser_mean) * (sold_ser - sold_ser_mean))) / np.sqrt(rating_ser_var * sold_ser_var)
print(f'Коэффициент корреляции: {pearson_corr}')

Коэффициент корреляции: 0.0006683352822727228


Статистика вычисляется по формуле

\begin{equation}
    t = \frac{\rho\sqrt{N - 2}}{\sqrt{1 - \rho^2}}
\end{equation}

In [70]:
print(f'Статистика: {(pearson_corr * np.sqrt(len(rating_ser) - 2)) / np.sqrt(1 - pearson_corr**2)}')

Статистика: 0.10166609977242959


Данная статистика подчиняется распределению Стьюдента с $N - 2$ степенями свободы

In [71]:
print(f'Критическое значение: {scp.stats.t.isf(0.05, df=(len(rating_ser) - 2))}')

Критическое значение: 1.6449194796102362


Статистика меньше критического значения, значит нет оснований для отклонения нулевой гипотезы

Вторым будет критерий Спирмена, статистика которого вычисляется аналогично предыдущей, но между рангами соответствующих векторов данных

\begin{equation}
    \rho = \frac{1}{N - 1} \frac{\sum_{i=1}^{N}{\left(R_i - \bar{R}\right)\left(S_i - \bar{S}\right)}}{\sqrt{s^2_R s^2_S}}
\end{equation}

где
* $R$, $S$ - ранги выборок $x$, $y$ размером $N$
* $\bar{R}$, $s^2_R$ и $\bar{S}$, $s^2_S$ - медиана и среднее соответственно $R$ и $S$

In [72]:
rating_ranks = scp.stats.rankdata(rating_ser)
sold_ranks = scp.stats.rankdata(sold_ser)

rating_ranks_mean = rating_ranks.mean()
sold_ranks_mean = sold_ranks.mean()

rating_ranks_var = rating_ranks.var()
sold_ranks_var = sold_ranks.var()

spearman_corr = (1 / (len(rating_ranks) - 1) * np.sum((rating_ranks - rating_ranks_mean) * (sold_ranks - sold_ranks_mean))) / np.sqrt(rating_ranks_var * sold_ranks_var)
print(f'Коэффициент корреляции: {spearman_corr}')

Коэффициент корреляции: 0.00047479636526410985


Статистика и критическое значение вычисляются аналогично

In [73]:
print(f'Статистика: {(spearman_corr * np.sqrt(len(rating_ranks) - 2)) / np.sqrt(1 - spearman_corr**2)}')

Статистика: 0.07222525966108365


In [74]:
print(f'Критическое значение: {scp.stats.t.isf(0.05, df=(len(rating_ranks) - 2))}')

Критическое значение: 1.6449194796102362


Снова, статистика меньше критического значения, поэтому нулевая гипотеза не отклоняется

Исходя из результатов используемых критериев, нулевая гипотеза принимается: рейтинг не зависит от цены